{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# LASSO Regression\n",
    "## Do Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-16T11:41:37.068693Z",
     "start_time": "2017-07-16T11:41:37.057924Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.linear_model import Lasso\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.pylab import rcParams\n",
    "import numpy\n",
    "#import mpld3\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import logging\n",
    "from scipy import stats\n",
    "import seaborn\n",
    "from decimal import Decimal\n",
    "\n",
    "\n",
    "FORMAT='%(asctime)s - %(name)s - %(levelname)s - %(message)s'\n",
    "logging.basicConfig(format=FORMAT)\n",
    "LOG=logging.getLogger()\n",
    "LOG.setLevel(logging.INFO)\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How to run this notebook\n",
    "To run the code in this notebook,\n",
    "\n",
    "    1) install [jupyter notebook](http://jupyter.org/install.html)\n",
    "    \n",
    "    2) open command prompt or terminal\n",
    "    \n",
    "    3) cd into directory containing this notebook and data\n",
    "    \n",
    "    4) run `$jupyter notebook` \n",
    "    \n",
    "    5) open this notebook\n",
    "    \n",
    "    6) change `dire` variable (below) to the directory where this document before running the code. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-16T11:41:41.515291Z",
     "start_time": "2017-07-16T11:41:41.481270Z"
    }
   },
   "outputs": [],
   "source": [
    "dire = r'C:\\Users\\Ciaran\\Documents\\Miscellaneous\\LassoRegression'\n",
    "dire = r'/home/b3053674/Documents/Miscellaneous/LassoRegression'\n",
    "## read data\n",
    "lasso_data_filename = os.path.join(dire, 'DataForLasso.csv')\n",
    "data = pandas.read_csv(lasso_data_filename,index_col=[0,1])\n",
    "data = data.transpose()\n",
    "\n",
    "\n",
    "## convert str column headers to int\n",
    "columns1 = data.columns.get_level_values(0)\n",
    "columns2 = data.columns.get_level_values(1)\n",
    "columns1 = [int(i) for i in columns1]\n",
    "columns2 = [int(i) for i in columns2]\n",
    "data.columns = [columns1,columns2]\n",
    "\n",
    "\n",
    "## get list of genes fro later\n",
    "genes = list(data.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lasso Regression\n",
    "I've isolated the TGFb treatment of the microarray data (BJ fibroblasts: 10ng/mL treatment over 180minutes). The  0 time point in this data was not treated with TGFb and is a `no-treatment` control. The data has 6 repeats, 8 time points (including the 0) and 217 differentially expressed genes (DEGs), as determined by LIMMA package in R. \n",
    "\n",
    "In this document I'm testing the LASSO regression technique to see how useful it can be with our data. I'll be using the JunB data as a response variable $Y$ and all other genes in the set of DEGs as predictor variables $X$\n",
    "\n",
    "A complication with our dataset is that we have time course data. For the analysis in the next cell, I've only considered the 180 minute data. \n",
    "\n",
    "### Summary\n",
    "Data: 6 repeats of TGFb treated microarray data at 180minutes post TGFb stimulation\n",
    "\n",
    "Dependent variable: JunB\n",
    "\n",
    "Independent variables: all other genes determined to be differentially expressed by LIMMA\n",
    "\n",
    "The [LassoCV](http://scikit-learn.org/stable/auto_examples/linear_model/plot_lasso_model_selection.html) class was used to estimated the best value of `alpha` by cross-validation\n",
    "\n",
    "### Functions\n",
    "The function called `lasso_for_timepoint` in the next cell performs a lasso regression analysis on the microarray DEG data at a single time point. The function `lasso_for_gene` iterates over time and uses the `lasso_for_timepoint` for each time point. \n",
    "\n",
    "### Overview of Algorithm\n",
    "\n",
    "Let: \n",
    "    \n",
    "    G = Number of Genes\n",
    "    R = Number of Repeats\n",
    "    T = Number of Time points\n",
    "    \n",
    "    y = Response Variable. All repeats of Gene g at time t. R x 1 vector: \n",
    "                        \n",
    "                        i.e if JunB == response variable                                 \n",
    "                                R  180min\n",
    "                                1   -\n",
    "                                2   -\n",
    "                                .   .\n",
    "                                6   -\n",
    "    \n",
    "    X = Predictor variables. All repeats of all genes != y at time t. RxG matrix\n",
    "    \n",
    "                        i.e. at t = 180min\n",
    "                            R    Gene_1    Gene_2    ...    Gene_n\n",
    "                            1       -         -      ...       -\n",
    "                            2       .                          . \n",
    "                            3       .                          .  \n",
    "                            4       .                          .\n",
    "                            5       .                          .\n",
    "                            6       .         -      ...      ...\n",
    "    for gene in Genes:\n",
    "        for t in T:\n",
    "            Get response variable data at time t\n",
    "            Get predictor variables at time t\n",
    "            Feed into sklearn.LassoCV (which estimates best alpha parameter for us)\n",
    "            Remove independent variables with 0 coefficient\n",
    "        return dataframe containing non-zero coefficients \n",
    "    Plot coefficients for each gene to pdf\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-16T11:35:07.238056Z",
     "start_time": "2017-07-16T11:35:07.200135Z"
    }
   },
   "outputs": [],
   "source": [
    "from sklearn import linear_model\n",
    "\n",
    "def lasso_for_timepoint(data, dependent_variable='JUNB', time_point=180, plot=False, max_iter = 10000):\n",
    "    \"\"\"\n",
    "    Perform lasso regression using sklearn.LassoCV on the microarray data for a single time point\n",
    "    \n",
    "    args:\n",
    "    \n",
    "        data: \n",
    "            Dataframe. columns = [repeats,time] multi-index\n",
    "            \n",
    "    kwargs:\n",
    "    \n",
    "        dependent_variable:\n",
    "            Which gene to consider the dependent variable. Default=JunB\n",
    "            \n",
    "        time_point:\n",
    "            Which time point to consider. Default=180min\n",
    "            \n",
    "        plot: \n",
    "            If True, plot data[dependent_variable].loc[time_point] Vs \n",
    "            any independent variables with non-zero coefficients\n",
    "            \n",
    "        max_iter: \n",
    "            Maximum number of iterations. Passed directly to LassoCV\n",
    "            \n",
    "    ==========\n",
    "    :returns: pandas.DataFrame containing any genes with non-zero coefficients \n",
    "    \"\"\"\n",
    "    if dependent_variable not in data.index:\n",
    "        raise Exception('{} not in data'.format(dependent_variable))\n",
    "    \n",
    "    if time_point not in list(set(data.columns.get_level_values(1))):\n",
    "        raise Exception('{} not in data'.format(time_point))\n",
    "\n",
    "    ## swap levels of column index\n",
    "    swap_data  = data.swaplevel(0,1,axis=1)\n",
    "\n",
    "    ## select just the time point of interest\n",
    "    time_point_data = swap_data[time_point].transpose()\n",
    "\n",
    "    ## isolate dependent variable data\n",
    "    dependent_variable_data = time_point_data[dependent_variable]\n",
    "    \n",
    "    ## Some genes have multiple probes. This interferes with the function so only\n",
    "    ## the first probe is considered. This code filters out extra probes\n",
    "    if isinstance(dependent_variable_data,pandas.core.frame.DataFrame):\n",
    "        dependent_variable_data = dependent_variable_data.iloc[:,0]\n",
    "        \n",
    "    ## Remove dependent_variable from independent variables\n",
    "    time_point_data.drop(dependent_variable,axis=1,inplace=True)\n",
    "\n",
    "    ## Fit LassoCV model\n",
    "    regr = linear_model.LassoCV(max_iter = max_iter, normalize = True)\n",
    "    regr.fit(time_point_data,dependent_variable_data)\n",
    "\n",
    "    ## get coefficients\n",
    "    coef = regr.coef_\n",
    "\n",
    "    ## Remove coefficients that Lasso set to 0\n",
    "    df = pandas.DataFrame(coef,index = time_point_data.columns)\n",
    "    df = df[df[0]!=0.0]\n",
    "    df.columns = [time_point]\n",
    "    df.index.name = 'Gene Symbol'\n",
    "\n",
    "    ## Get list of genes\n",
    "    genes = data.keys()\n",
    "    ## Plot\n",
    "    if plot:\n",
    "        for gene in genes:\n",
    "            if gene in df.index:\n",
    "                plt.figure()\n",
    "                plt.plot(dependent_variable_data, time_point_data[gene],'o')\n",
    "                plt.title(' {} Vs {}'.format(dependent_variable, gene))\n",
    "                plt.ylabel('{}@{}min-post-tgf'.format(gene, time_point))\n",
    "                plt.xlabel('{}@{}min-post-tgf'.format(dependent_variable, time_point))\n",
    "                \n",
    "    return df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-16T11:35:22.608231Z",
     "start_time": "2017-07-16T11:35:22.585128Z"
    }
   },
   "outputs": [],
   "source": [
    "def lasso_for_gene(data, dependent_variable='JUNB', plot=True, max_iter = 1e4):\n",
    "    \"\"\"\n",
    "    Run a lasso regression for each time point in the gene data\n",
    "    \n",
    "    args:\n",
    "        data for conducting regression on (see `data` variable)\n",
    "    \n",
    "    kwargs:\n",
    "        dependent_variable. Response variable. Default = JunB\n",
    "        \n",
    "        plot:\n",
    "            plot a spy plot. colours indicate non-zero coefficient. Default = True\n",
    "            \n",
    "        max_iter:\n",
    "            parameter passed on to sklearn.LassoCV. Maximum number of iterations for fitting algorithm. \n",
    "    \"\"\"\n",
    "\n",
    "    ## iterate over lasso_for_timepoint for each time point, save results as dict\n",
    "    dependent_variable_coefficients =  {time: lasso_for_timepoint(data,\n",
    "                              dependent_variable=dependent_variable,\n",
    "                              time_point = time,\n",
    "                              plot = False,\n",
    "                              max_iter=max_iter) for time in list(set(data.columns.get_level_values(1)))}\n",
    "\n",
    "    ## concatonate the results\n",
    "#     print dependent_variable_coefficients\n",
    "    df = pandas.concat(dependent_variable_coefficients, axis=1)    \n",
    "\n",
    "    # fix columns\n",
    "    df.columns = df.columns.droplevel(0)\n",
    "    \n",
    "    ## plot as spy plot\n",
    "    if plot:\n",
    "        x_labels = ['Gene'] + list(df.columns)\n",
    "        y_labels = list(df.index)\n",
    "        fig, ax = plt.subplots()\n",
    "        ax.spy(df, aspect='auto')\n",
    "        ax.set_xticklabels(x_labels)\n",
    "        plt.title(dependent_variable)\n",
    "\n",
    "        ## Fix y ticks\n",
    "        highy, lowy = ax.get_ylim()\n",
    "        plt.yticks(numpy.arange(lowy+0.5,highy+0.5, 1.0))\n",
    "        #ax.set_yticks = y_ticks\n",
    "        ax.set_yticklabels(y_labels)\n",
    "        plt.xlabel('Time(min)')\n",
    "    return df\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-16T11:38:38.550631Z",
     "start_time": "2017-07-16T11:35:24.992764Z"
    }
   },
   "outputs": [],
   "source": [
    "## Perform a lasso regression analysis per time point for gene in set of genes\n",
    "## Save to pdf file in current directory if SAVE set to True\n",
    "SAVE = False\n",
    "with PdfPages('LassoCVOutput.pdf') as pdf:\n",
    "    for gene in list(set(genes)):\n",
    "        df = lasso_for_gene(data=data, dependent_variable=gene, plot=True, max_iter=1e5)\n",
    "        if SAVE:\n",
    "            pdf.savefig(dpi=300,bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Multi-task Lasso\n",
    "A problem with the previous approach was that LassoCV does not consider multivariate dependentent or independentant variables. The [multi-task Lasso](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.MultiTaskLasso.html) however is capable of jointly considering as many dependentent variables as you like: i.e. as many time points for a gene as we like. Its become apparent however that the multi-task lasso cannot jointly consider the time points of the independent variable. \n",
    "\n",
    "\n",
    "Let: \n",
    "    \n",
    "    G = Number of Genes\n",
    "    R = Number of Repeats\n",
    "    T = Number of Time points\n",
    "    \n",
    "    Y = Response Variable. All repeats of Gene g at time t. R x T matrix: \n",
    "                        \n",
    "                        i.e if JunB == response variable    \n",
    "                        \n",
    "                                R  0    15    30    ...    180min\n",
    "                                1  -                          .\n",
    "                                2  -     .                    .\n",
    "                                3  .           .              .\n",
    "                                4  .                 .        .\n",
    "                                5  .                          .\n",
    "                                6  .     .     .    ...       .       \n",
    "    \n",
    "    X = Predictor variables. All repeats of all genes != y at time t. RxG matrix\n",
    "    \n",
    "                        i.e. at t = 180min\n",
    "                            R    Gene_1    Gene_2    ...    Gene_n\n",
    "                            1       -         -      ...       -\n",
    "                            2       .                          . \n",
    "                            3       .                          .  \n",
    "                            4       .                          .\n",
    "                            5       .                          .\n",
    "                            6       .         -      ...      ...\n",
    "    for t in T:\n",
    "        Get response variable data at time t\n",
    "        Get predictor variables at time t\n",
    "        Feed into sklearn.MultitaskLasso (but now we need to provide alpha parameter ourselves)\n",
    "        Remove independent variables with 0 coefficient\n",
    "    return dataframe containing non-zero coefficients \n",
    "    \n",
    "    Plot to pdf\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-16T11:33:01.860732Z",
     "start_time": "2017-07-16T11:32:56.968188Z"
    }
   },
   "outputs": [],
   "source": [
    "from sklearn.linear_model import MultiTaskLasso\n",
    "\n",
    "\n",
    "def multi_task_lasso(data, dependent_variable='JUNB', alpha=0.05, max_iter=1e5, plot=True,normalize=True):\n",
    "    \"\"\"\n",
    "    \n",
    "    \"\"\"\n",
    "    LOG.info('Computing lasso coefficients for: {} at alpha: {}'.format(dependent_variable,alpha))\n",
    "    data = data.transpose()\n",
    "    ## isolate the dependent variable\n",
    "    \n",
    "    Y = data[dependent_variable].unstack()\n",
    "    \n",
    "    ## get independent variable DataFrame\n",
    "    Xt = {label:df for label,df in data.groupby(level=1)} #for t in set of time points\n",
    "    \n",
    "    ## Fit multitask lasso model for each time point in independent variables\n",
    "    ## extract the coefficients and save to pandas dataframe with appropriate columns/index\n",
    "    coef_dct = {time: pandas.DataFrame(\n",
    "                            MultiTaskLasso(\n",
    "                                alpha=alpha, max_iter=max_iter, normalize=normalize).fit(\n",
    "                                Xt[time],Y).coef_,\n",
    "                            columns = data.columns,\n",
    "                            index = sorted(list(set(data.index.get_level_values(1))))) for time in Xt}\n",
    "\n",
    "    ## remove independent variables with zeros\n",
    "    coef_dct_no_zeros = {time: coef_dct[time][coef_dct[time]!=0.0].dropna(how='all',axis=1) for time in coef_dct.keys()}\n",
    "    \n",
    "    \n",
    "    if plot:\n",
    "        fig, ax = plt.subplots(nrows=8, ncols=1,figsize=(30,20))\n",
    "        \n",
    "\n",
    "        for t in range(1,len(coef_dct_no_zeros.keys())+1  ):\n",
    "            keys = sorted(coef_dct_no_zeros.keys())\n",
    "            plt.subplot(8,1,t)\n",
    "            seaborn.heatmap(data=coef_dct_no_zeros[keys[t-1]], \n",
    "                            cbar=True,\n",
    "                            cbar_kws={'label': 'colorbar title'})\n",
    "            plt.ylabel('X at {}min\\nYtasks=Time(min)'.format(keys[t-1]), fontsize=12)\n",
    "\n",
    "\n",
    "        plt.suptitle('Coefficients from multi-task LASSO regression (alpha={}\\nfor response varibale Y={}'.format(\n",
    "            alpha,dependent_variable), fontsize=24)\n",
    "\n",
    "\n",
    "        plt.tight_layout()\n",
    "        plt.subplots_adjust(top=0.95)\n",
    "#         plt.savefig('{}.jpeg'.format(dependent_variable), dpi=300, bbox_inches='tight')\n",
    "    return fig,coef_dct_no_zeros\n",
    "\n",
    "gene_index=50\n",
    "gene = genes[gene_index]\n",
    "gene = 'JUNB'\n",
    "\n",
    "multi_task_lasso(data,dependent_variable=gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Iteration for all genes\n",
    "The following code iterates over the DEG gene list with the multi-task lasso function written in the previous cell. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "\n",
    "def iterate_for_all_genes(alpha):\n",
    "    with PdfPages('MultiTaskLassoOutputAlpha{}.pdf'.format(alpha)) as pdf:\n",
    "        for gene in genes:\n",
    "            multi_task_lasso(data,dependent_variable=gene,alpha=alpha)\n",
    "            pdf.savefig(dpi=300, bbox_inches='tight')\n",
    "            \n",
    "\n",
    "# alpha = [0.5,0.7]\n",
    "iterate_for_all_genes(0.01)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Second round of Lasso \n",
    "After some discussions the idea of including all time points in the regression has been suggested. Average the data, get variances at each time point and perform regression on full time courses using variances as weights. In essence we'll neglect the time element and consider each gene an individual category (instead of each time of each gene). \n",
    "\n",
    "## Average the data and get variances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-16T11:38:38.691587Z",
     "start_time": "2017-07-16T11:38:38.676491Z"
    }
   },
   "outputs": [],
   "source": [
    "summary = data.groupby(level=1,axis=1).aggregate(numpy.mean)#, numpy.std)\n",
    "std = data.groupby(level=1,axis=1).aggregate(numpy.std)\n",
    "\n",
    "'''\n",
    "Sanity check:\n",
    "\n",
    "For Gene: ABHD17C\n",
    "\n",
    "Repeat    0 timepoint value\n",
    "1            7.996778    \n",
    "2            8.024219\n",
    "3            8.078087\n",
    "4            8.127900\n",
    "5            8.131485\n",
    "6            8.050835\n",
    "    \n",
    "'''\n",
    "abhd17c = [7.996778, 8.024219, 8.078087,\n",
    "           8.127900, 8.131485, 8.050835]\n",
    "\n",
    "abhd17c_average = sum(abhd17c)/len(abhd17c)\n",
    "abhd17c_std = numpy.std(abhd17c)\n",
    "\n",
    "LOG.info('Testing whether manually computer mean == pandas computer mean --> {} == {}'.format(\n",
    "    round(abhd17c_average,6), round(summary.loc['ABHD17C'][0], 6) ))\n",
    "\n",
    "LOG.info('Testing whether manually computer std == pandas computer std --> {} == {}'.format(\n",
    "    round(abhd17c_std,6), round(std.loc['ABHD17C'][0], 6) ))  ##not exactly correct. I think this is floatig point error however\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let: \n",
    "    \n",
    "    G = Number of Genes\n",
    "    R = Number of Repeats\n",
    "    T = Number of Time points\n",
    "    \n",
    "    Y = Response Variable. Averaged repeats for each time point of gene g\n",
    "                        \n",
    "                        i.e    \n",
    "                        \n",
    "                                t     g\n",
    "                                0     -\n",
    "                                15    -\n",
    "                                30    -\n",
    "                                60    -\n",
    "                                90    -\n",
    "                                120   -\n",
    "                                150   -\n",
    "                                180   -\n",
    "    \n",
    "    X = Predictor variables. All repeats of all genes != y at time t. RxG matrix\n",
    "    \n",
    "                        i.e.\n",
    "                        \n",
    "                                t     g1    g2    gn\n",
    "                                0     -     -     -\n",
    "                                15    -     -     -\n",
    "                                30    -     -     -\n",
    "                                60    -     -     -\n",
    "                                90    -     -     -\n",
    "                                120   -     -     -\n",
    "                                150   -     -     -\n",
    "                                180   -     -     -\n",
    "                            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-16T11:41:57.281504Z",
     "start_time": "2017-07-16T11:41:57.131281Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from sklearn import linear_model\n",
    "\n",
    "\n",
    "def ElasticNetRegression(data,\n",
    "                         dependent_variable = 'JUNB',\n",
    "                         max_iter = 1e5,\n",
    "                         filename = 'RegressionAnalysisGraphs',\n",
    "                         plot_heatmap=False,\n",
    "                         plot_regression_square=False,\n",
    "                         plot_regression_singles=False,\n",
    "                         save=False):\n",
    "    \"\"\"\n",
    "    args:\n",
    "        data to perform regression on. Genes are rownames and each row\n",
    "        corresponds to a single genes expression values. Colnames are irrelevant\n",
    "        \n",
    "    kwargs:\n",
    "        dependent_variable:\n",
    "            variable which takes on the y (predicted) role\n",
    "            \n",
    "        max_iter:\n",
    "            numeric. Passed on to sklearn.linear_model.<function>\n",
    "            \n",
    "        plot_heatmap:\n",
    "            Boolean. Plot regression coefficients, pearsons coef. and pearsons coef p-value\n",
    "            as heat maps\n",
    "            \n",
    "        plot_regression_square:\n",
    "            Boolean. Filter data to only include genes which has non-zero coefficient, \n",
    "            determined by sklearn.linearmodel then plot each gene against the dependent variable\n",
    "            as a square\n",
    "            \n",
    "        plot_regression_single:\n",
    "            Boolean. Same as plot_regression_square but plot one graph per plot. \n",
    "\n",
    "        folder:\n",
    "            name of folder to save graphs to\n",
    "        \n",
    "        save:\n",
    "            boolean. save to file\n",
    "    \"\"\"\n",
    "\n",
    "    \n",
    "    ## check dep variable is in genes\n",
    "    genes = list(data.index)\n",
    "    if dependent_variable not in genes:\n",
    "        raise Exception('{} not in genes'.format(dependent_variable))\n",
    "    \n",
    "    ## Format data for input into regression model\n",
    "    data = data.transpose()\n",
    "    y = pandas.DataFrame(data[dependent_variable])\n",
    "    \n",
    "    # do regression\n",
    "    reg = function(normalize=True, max_iter=max_iter)\n",
    "    reg.fit(data, y)\n",
    "    \n",
    "    ## get coefficients\n",
    "    df = pandas.DataFrame(reg.coef_, columns=data.columns, index=['Regression Coef']).transpose()\n",
    "    \n",
    "    ## extract non-zero coefficients\n",
    "    df = df[df['Regression Coef']!=0]\n",
    "    df = df.sort_values(by='Regression Coef', ascending=False)\n",
    "    \n",
    "    ## calculate pearsons correlation coef\n",
    "    genes = list(df.index)\n",
    "    r=[]\n",
    "    for gene in genes:\n",
    "        coef = stats.pearsonr(data[gene],data[dependent_variable])\n",
    "        r.append(coef)\n",
    "    r_df = pandas.DataFrame(r, index=genes, columns = ['Pearsons Coef', 'p-value'])\n",
    "    \n",
    "    ## merge pearsons correlation and p_value with regression coefficients \n",
    "    df=df.sort_index()\n",
    "    r_df=r_df.sort_index()    \n",
    "    heat_map_df = pandas.merge(df, r_df, left_index=True, right_index=True)\n",
    "    \n",
    "    ## calculate mutual information as well\n",
    "    \n",
    "    ## collect figures for plotting later\n",
    "    figs=[]\n",
    "    heat_map_df = heat_map_df.sort_values(by='p-value')\n",
    "    seaborn.set_context(context='poster', font_scale=1)\n",
    "    if plot_heatmap:\n",
    "        LOG.info('\\tplotting heatmap for {}'.format(dependent_variable))\n",
    "        fig=plt.figure(figsize=(40,0.5*heat_map_df.shape[0]))\n",
    "        for i in range(heat_map_df.shape[1]):\n",
    "            plt.subplot(1,3,i+1)\n",
    "            seaborn.heatmap( pandas.DataFrame(heat_map_df[heat_map_df.columns[i]]))\n",
    "            plt.suptitle('Dependent Variable: {}'.format(dependent_variable) )\n",
    "        figs.append(fig)\n",
    "\n",
    "            #         if save:\n",
    "#             pdf.savefig(bbox_inches='tight', dpi=300)\n",
    "\n",
    "\n",
    "    ## plot regressions as square\n",
    "    num_non_zero_genes=len(list(df.index))\n",
    "    num_non_zero_root = numpy.ceil(numpy.sqrt(num_non_zero_genes))\n",
    "    if plot_regression_square:\n",
    "        data.index.name='time'\n",
    "        data=data.reset_index()\n",
    "        fig=plt.figure()\n",
    "        for i in range(len(list(df.index))):\n",
    "            gene = list(df.index)[i]\n",
    "            LOG.info('\\tplotting regplot (square) for {}'.format(gene))\n",
    "            \n",
    "            plt.subplot(num_non_zero_root,num_non_zero_root,i+1)\n",
    "            seaborn.regplot(x=data[gene], y=data[dependent_variable], data=data )\n",
    "            plt.ylabel('')\n",
    "            plt.xlabel('')\n",
    "            plt.title(gene)\n",
    "            plt.suptitle('Regression plots between {} and genes with non-zero \\nelastic net coefficient'.format(dependent_variable))\n",
    "        plt.tight_layout()\n",
    "        plt.subplots_adjust(top=0.85)\n",
    "        figs.append(fig)\n",
    "#         if save:\n",
    "#             pdf.savefig(bbox_inches='tight', dpi=300)\n",
    "\n",
    "\n",
    "    ## plot regression as single\n",
    "    if plot_regression_singles:\n",
    "        data.index.name='time'\n",
    "        for gene in genes:\n",
    "            LOG.info('\\tPlotting regplot (single) for {}'.format(gene))\n",
    "            plt.figure()\n",
    "#             seaborn.lmplot(x=gene, y=dependent_variable, data=data)\n",
    "            seaborn.regplot(x=data[gene], y=data[dependent_variable], data=data)\n",
    "            plt.title('Elastic Net Regression, {} Vs {}. \\n\\\n",
    "Gene Expression (FC) of BJ fibroblast Reponse to TGFb\\n\\\n",
    "ElasticNetRegCoef.={}, Pearsons: {}, p-val: {}'.format(\n",
    "                gene, dependent_variable,\n",
    "                round(heat_map_df['Regression Coef'][gene], 4),\n",
    "                round(heat_map_df['Pearsons Coef'][gene], 4),\n",
    "                round(heat_map_df['p-value'][gene])))\n",
    "#         if save:\n",
    "#             pdf.savefig(bbox_inches='tight', dpi=300)\n",
    "    return figs, df, heat_map_df\n",
    "\n",
    "\n",
<<<<<<< HEAD
    "figs = ElasticNetRegression(summary, 'JUNB', plot_heatmap=True, plot_regression_square=False,\n",
    "                         plot_regression_singles=False)"
=======
    "figs = ElasticNetRegression(summary, 'JUNB', plot_heatmap=True, plot_regression_square=True,\n",
    "                         plot_regression_singles=True, %save=True)"
>>>>>>> f25ba09930ccdb13c9c226f9a46b518e37285b6e
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-09T08:24:12.294000Z",
     "start_time": "2017-07-09T08:23:33.308000Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "##copy frame and relabel columns to include all repeats individually in the regression\n",
    "data_with_repeats = data.copy()\n",
    "\n",
    "## informative labels which can be used by scatters\n",
    "data_with_repeats.columns = list(set(data.columns.get_level_values(1)))*max(data.columns.get_level_values(0))\n",
    "\n",
    "figs = ElasticNetRegression(data_with_repeats, 'JUNB', plot_heatmap=True, plot_regression_square=True,\n",
    "                         plot_regression_singles=True)\n",
    "def run_all():\n",
    "    SAVE=True\n",
    "    with PdfPages(os.path.join(os.getcwd(), 'RegressionAnalysisOutput.pdf')) as pdf:\n",
    "        for gene in sorted(genes):\n",
    "            LOG.info('Calculating ElasticNet regression for {} as dependent variable'.format(gene))\n",
    "            figs = ElasticNetRegression(summary, gene, plot_heatmap=True, plot_regression_square=False,\n",
    "                         plot_regression_singles=False)\n",
    "            if SAVE:\n",
    "                [pdf.savefig(fig, bbox_inches='tight', dpi=300) for fig in figs[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-15T13:30:33.622001Z",
     "start_time": "2017-07-15T13:30:33.619516Z"
    }
   },
   "source": [
    "# Mutual Information "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-15T13:31:46.196694Z",
     "start_time": "2017-07-15T13:31:46.133981Z"
    }
   },
   "source": [
    "MU can be used in the sklearn package. Sklearn will also do stability selection. look into this!!1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandas.tools.plotting import scatter_matrix\n",
    "scatter_matrix(data, alpha=0.2, figsize=(6, 6), diagonal='kde')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "227px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_position": {
    "height": "907px",
    "left": "0px",
    "right": "1643px",
    "top": "130px",
    "width": "212px"
   },
   "toc_section_display": "block",
   "toc_window_display": true,
   "widenNotebook": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
